### PR TITLE
enable optionally specifying a subset of object classes

### PR DESCRIPTION
**Pull request by definitelyuncertain**
Issue: #11972 

This feature allows the user to _optionally_ specify a subset of object classes to include during training or validation through the data YAML. From the included example in data/coco128LabelSubset.yaml :

```[yaml]
include_names:
  - bicycle
  - motorcycle
  - car
  - bus
  - train
  - truck
```

Including the above section in the data YAML would cause only the classes mentioned there to be used for training, and also re-number the classes in the order in which they are specified. So the trained model would have class 0 as `bicycle`, class 1 as `motorcycle`, and so on. If the above section is not specified, all classes available in the dataset are used as usual.

The only code changes are to check for the above section in the data YAML in the train and validation scripts, and pass it into the data loader. There, it re-numbers the included subset of classes and applies the new class numbers while loading the labels.

Tested training and validation with YOLOv5 on COCO-128 (data/coco128LabelSubset.yaml) and a custom dataset.

### Summary
📄🎛️🚗

This pull request adds a new feature to train and validate YOLOv5 models on a subset of classes defined by the user in the YAML file. It modifies the `train.py`, `val.py`, and `utils/dataloaders.py` files to handle the class filtering logic. It also provides an example YAML file `data/coco128LabelSubset.yaml` that uses a subset of the COCO128 dataset.

> _Sing, O Muse, of the skillful code that modified YOLOv5_
> _To train and test on custom classes, a subset of COCO128_
> _With `include_names` in YAML, they filtered the data well_
> _Like Zeus who thunders from Olympus, striking down the rebel_

### Walkthrough
*  Add a new YAML file `coco128LabelSubset.yaml` to define a custom class subset of the COCO128 dataset ([link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-8bb1b1f74adbaf3a2301e876faab735e5a10f14f1ec09aa18811a2e7f1920e6fR1-R109))
*  Add a new parameter `include_classes` to the `create_dataloader` function and the `LoadImagesAndLabels` class to filter the labels and segments according to the class subset ([link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-03f58e512c7bd0d2f1cf615a4768242b2aeb8272d7837c076e1bf43f92fe2216L119-R120), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-03f58e512c7bd0d2f1cf615a4768242b2aeb8272d7837c076e1bf43f92fe2216L136-R138), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-03f58e512c7bd0d2f1cf615a4768242b2aeb8272d7837c076e1bf43f92fe2216L452-R455), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-03f58e512c7bd0d2f1cf615a4768242b2aeb8272d7837c076e1bf43f92fe2216L533-R559))
*  Add a new variable `include_classes` to the `train` and `run` functions in `train.py` and `val.py` to obtain the class subset from the YAML file and pass it to the `create_dataloader` function ([link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R123-R132), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95L210-R221), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95L228-R240), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-9d519d95d7e57936997ef1b72f9cb5ade1003df487f612dfaf5427bda532830fR166-R176), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-9d519d95d7e57936997ef1b72f9cb5ade1003df487f612dfaf5427bda532830fL183-R195))
*  Add log messages to inform the user about the included classes in `train.py` and `val.py` ([link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R123-R132), [link](https://github.com/ultralytics/yolov5/pull/11973/files?diff=unified&w=0#diff-9d519d95d7e57936997ef1b72f9cb5ade1003df487f612dfaf5427bda532830fR166-R176))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introducing class filtering for training and validation in YOLOv5.

### 📊 Key Changes
- 🔧 New `include_names` section in the dataset configuration file to specify a subset of classes for training.
- 🛠 Modified `train.py`, `val.py`, and `utils/dataloaders.py` to handle class filtering based on `include_names`.
- ➕ Added logic to map user-specified subset class names to their corresponding indices in the dataset.
- 🔄 Updates to `LoadImagesAndLabels` to use optional `include_classes`, updating labels with new class numbers.
- ✂️ Filtering of labels and segments to only include specified classes, and re-indexing classes for training/validation consistency.

### 🎯 Purpose & Impact
- 🏷 Allows users to train models on a subset of classes, reducing complexity and potentially improving performance for specific tasks.
- 🎛 Users gain more control over their datasets and can fine-tune models with greater precision.
- 🚀 This update can make YOLOv5 more flexible and efficient for custom detection tasks with fewer classes.

### CODE REVIEW

**Comment by definitelyuncertain** [comment]
@glenn-jocher any interest in this PR for issue #11972 ?

Also, is there a CLA I need to sign? I couldn't see anything about one in the contribution instructions.

**Comment by glenn-jocher** [comment]
@definitelyuncertain hi there! Thank you for your contribution and interest in addressing issue #11972. We appreciate your effort in adding the feature to optionally specify a subset of object classes.

Regarding your question about the Contributor License Agreement (CLA), we don't currently have one in place for YOLOv5. Feel free to proceed without it for now.

Please let us know if you have any further questions or need any assistance with your contribution. We're here to help!

**Comment by definitelyuncertain** [comment]
@glenn-jocher The checks are all passing, so feel free to review and let me know if there's something I need to change.

**Comment by glenn-jocher** [comment]
@definitelyuncertain thanks for letting us know! We'll review your changes and provide feedback soon. Your contribution is greatly appreciated.

**Comment by WangQvQ** [comment]
@definitelyuncertain Thanks for your contribution, but I have some questions about the logic. First of all, when I use COCO128 data set, I only train one category, which will also calculate the classification loss. Moreover, when I print mAP after training, the category I do not want to train will still be printed.
image

image

image

**Comment by WangQvQ** [comment]
If I want to do this, I think I can just modify the include_class = [] in utils/ dataloads.py. If I only want to train the 0 class, I just have my include_class = [0]

I have tested the reasonableness of this practice, would you please have a look at it? @glenn-jocher 
image

**Comment by glenn-jocher** [comment]
@WangQvQ, thank you for your suggestion. Modifying the `include_class` list in `utils/dataloaders.py` can indeed be a viable solution if you want to train only a specific class, such as class 0. However, please note that this approach will still calculate the classification loss for all the classes in the dataset, even if they are not being trained.

In terms of the mAP metric, it is expected that all classes present in the dataset will be evaluated and printed, regardless of whether they were trained or not. This is because mAP measures the overall performance of the model across all classes in the dataset.

If you have tested this approach and found it to meet your needs, it is certainly worth considering. However, please keep in mind the nuances mentioned above. Let us know if you have any further questions or need additional assistance.

**Comment by WangQvQ** [comment]
@glenn-jocher Thank you very much for your reply, which is of great value to me！

**Comment by glenn-jocher** [comment]
@WangQvQ you're welcome! I'm glad my reply was helpful to you. If you have any further questions or need any more assistance, please don't hesitate to ask. We're here to support you!

**Comment by definitelyuncertain** [comment]
Hi @WangQvQ  , thanks for testing! However, I'm not able to reproduce what's happening in your first output screenshot. When is it printing the mAP for whole list of classes?

As for the second output screenshot, things appear to be working as intended. It says `all` but it also says `Including classes [1]` at each epoch, indicating that it is only calculating the mAP for the included classes.

Also, it is better to be able to specify which classes you want to train in the config file rather than having to modify the source code every time you wanted to include a different subset of classes.

**Comment by glenn-jocher** [comment]
Hi @definitelyuncertain, thank you for sharing your findings and testing the behavior of the YOLOv5 model.

Regarding your first output screenshot, I have not been able to reproduce the issue you mentioned where the mAP is being printed for the whole list of classes. Could you please provide more details about the specific scenario or code snippet that triggers this behavior? 

Regarding your second output screenshot, the behavior appears to be working as intended. The output states `all`, but it also shows the inclusion of specific classes at each epoch, such as `Including classes [1]`, indicating that the mAP calculation is indeed restricted to the specified classes.

I agree with your suggestion of specifying which classes to train in the config file rather than modifying the source code each time to include different class subsets. This approach adds flexibility and maintainability to the training process.

If you have any further details or questions, please let us know. We appreciate your valuable input and efforts in contributing to YOLOv5!
